### PR TITLE
Private chat mode

### PR DESCRIPTION
**Pull request by AKMDM**
This adds a new chat mode to Zandronum, allowing the player (or the server) to send chat messages privately to another player. Nobody else can see these messages and they won't be logged into the server's chat log, unless the server is sending or receiving a private message.

### CODE REVIEW

File: src/c_console.cpp [hunk 1]
```diff
@@ -285,6 +286,12 @@ CUSTOM_CVAR (Int, msg4color, 0, CVAR_ARCHIVE)
 	setmsgcolor (4, self);
 }
 
+CUSTOM_CVAR (Int, msg5color, 21, CVAR_ARCHIVE)
+{
+	// [AK] Private chat message color.
```
Comment (TorrSamaho) [inline src/c_console.cpp]: Please move this comment to where the delta to GZDoom starts, i.e. before `CUSTOM_CVAR (Int, msg5color ....`

File: src/cl_commands.cpp [hunk 2]
```diff
@@ -282,6 +282,22 @@ void CLIENTCOMMANDS_Say( ULONG ulMode, const char *pszString )
 	CLIENT_GetLocalBuffer( )->ByteStream.WriteString( chatstring );
 }
 
+//*****************************************************************************
+//
+void CLIENTCOMMANDS_PrivateSay( ULONG ulPlayer, const char *pszString )
```
Comment (TorrSamaho) [inline src/cl_commands.cpp]: This function shares a lot of code with `CLIENTCOMMANDS_Say`. I think it would be better if `CLIENTCOMMANDS_Say` is extended to cover the case that `ulMode` is `CHATMODE_PRIVATE_SEND`. For this it would need an additional optional argument `ulPlayer` that defaults to `MAXPLAYERS`.

File: src/doomtype.h [hunk 3]
```diff
@@ -151,7 +153,8 @@ enum
 #define PRINT_HIGH				2				// critical messages
 #define PRINT_CHAT				3				// chat messages
 #define PRINT_TEAMCHAT			4				// chat messages from a teammate
-#define PRINT_LOG				5				// only to logfile
+#define PRINT_PRIVATECHAT		5				// private chat messages from another player
+#define PRINT_LOG				6				// only to logfile
```
Comment (TorrSamaho) [inline src/doomtype.h:159]: This delta to GZDoom code should be marked with a comment. I'd also not change `PRINT_LOG`, since that is from GZDoom. Or is it necessary that `PRINT_PRIVATECHAT` is `PRINT_TEAMCHAT+1`?

Comment (AKMDM) [reply src/doomtype.h:159]: I figured private chat messages would have greater importance than team chat messages, so it should be one level higher. Would changing `PRINT_PRIVATECHAT` so that it matched `PRINT_LOG 5` be acceptable?

Comment (TorrSamaho) [reply src/doomtype.h:159]: Ah, I see. In that case it's fine to raise `PRINT_LOG` to `6`. Just leave comments to mark that you added `PRINT_PRIVATECHAT` and that you raised `PRINT_LOG` to make room.

File: src/sv_commands.cpp [hunk 4]
```diff
@@ -1061,6 +1061,33 @@ void SERVERCOMMANDS_PlayerSay( ULONG ulPlayer, const char *pszString, ULONG ulMo
...
+	// [AK] First send the command to the player receiving the message.
+	if (( ulReceiver != MAXPLAYERS ) && ( !bForbidChatToPlayers || players[ulReceiver].bSpectating ))
+	{
+		command.SetPlayerNumber( ulSender );
+		command.SetMode( CHATMODE_PRIVATE_RECEIVE );
+		command.sendCommandToClients( ulReceiver, SVCF_ONLYTHISCLIENT );
+	}
+
+	// [AK] Next, send the command back to the player who sent the message.
+	if ( ulSender != MAXPLAYERS )
```
Comment (TorrSamaho) [inline src/sv_commands.cpp:1090]: This will happen even if the message was not send due to `(( ulReceiver != MAXPLAYERS ) && ( !bForbidChatToPlayers || players[ulReceiver].bSpectating ))` being false. I think in that case, the sender should receive an error.

**Comment by AKMDM** [comment]
I'm finishing up on these adjustments, if that's everything. I should have a new commit ready very soon.

EDIT: Just committed the adjustments, I hope these are okay. One more thing, I added extra code to allow the client to cycle backwards through the player list if they're pressing TAB + SHIFT. Otherwise, it cycles forward with just TAB.

**Comment by TorrSamaho** [comment]
> I'm finishing up on these adjustments, if that's everything. I should have a new commit ready very soon.


Yes, I'm done checking and commented on everything I noticed.

File: src/chat.cpp [hunk 5]
```diff
@@ -609,17 +609,23 @@ bool CHAT_Input( event_t *pEvent )
 			{
 				if (( g_ulChatMode == CHATMODE_PRIVATE_SEND ) && ( g_ulChatPlayer != MAXPLAYERS ))
 				{
-					ULONG oldPlayer = g_ulChatPlayer;
+					int tempPlayer = g_ulChatPlayer;
+					int direction = ( pEvent->data3 & GKM_SHIFT ) ? -1 : 1;
```
Comment (TorrSamaho) [inline src/chat.cpp:614]: Are these changes here related to any of the points I brought up?  At first glance, it doesn't look like this to me. I also just realize that github seems to have hid the changes to chat.cpp when I made the review.

Comment (AKMDM) [reply src/chat.cpp:614]: This particular change isn't related to what you mentioned last week, it's just something I forgot to add when I originally made the pull request.

Originally, the client could only move forwards through the player list if they wanted to change who to send the private message to, whenever they pressed TAB. Now, they can move backwards if they press TAB + SHIFT at the same time. This is useful in case the client were to accidentally pass the player they intend to privately chat with and need to go backward/forward.

File: src/chat.cpp [hunk 6]
```diff
@@ -594,7 +607,30 @@ bool CHAT_Input( event_t *pEvent )
 			}
 			else if ( pEvent->data1 == '\t' )
 			{
-				g_ChatBuffer.TabComplete();
+				if (( g_ulChatMode == CHATMODE_PRIVATE_SEND ) && ( g_ulChatPlayer != MAXPLAYERS ))
```
Comment (TorrSamaho) [inline src/chat.cpp]: I think it would be safer to also check `g_ulChatPlayer != consoleplayer` and `SERVER_CalcNumPlayers() >= 2`. In case both are false, the while loop won't terminate.

File: src/chat.cpp (hunk 5 above)
Comment (TorrSamaho) [reply src/chat.cpp:614]: I see, sounds like a nice addition.

File: src/chat.cpp [hunk 7]
```diff
@@ -594,7 +607,30 @@ bool CHAT_Input( event_t *pEvent )
...
+
+						if ( tempPlayer < 0 )
+							tempPlayer = MAXPLAYERS - 1;
+						else if ( tempPlayer >= MAXPLAYERS )
+							tempPlayer = 0;
+
+						if ( tempPlayer == g_ulChatPlayer )
+							break;
+					}
+					while (( PLAYER_IsValidPlayer( tempPlayer ) == false ) || ( tempPlayer == consoleplayer ));
```
Comment (TorrSamaho) [inline src/chat.cpp]: Perhaps you want to skip bots? In that case, you need to check `SERVER_CountPlayers( false ) >= 2` instead of `SERVER_CalcNumPlayers() >= 2` in the comment above.

EDIT: I just noticed that we cannot select the server with this. Wouldn't it be more user friendly if we'd also allow the server to by cycled through? This way one also doesn't need to count the players to prevent an infinite loop.

Comment (AKMDM) [reply src/chat.cpp]: Maybe so. Perhaps the client should also only be allowed to use the private chat if the network state is `NETSTATE_CLIENT`? There might not be any real reason for the player to send private messages in a singleplayer game.

File: src/chat.cpp [hunk 8]
```diff
@@ -1131,6 +1225,152 @@ CCMD( say_team )
...
+	}
+}
+
+CCMD( sayto )
+{
+	ULONG ulPlayer = 0;
+
+	if ( argv.argc() >= 2 )
+	{
+		if ( stricmp( argv[1], "Server" ) == 0 )
```
Comment (TorrSamaho) [inline src/chat.cpp:1387]: So if a player has the name "Server", this player will not be reachable by `sayto`, right? I don't think that this is really a problem, but one should just be aware of this.

File: src/chat.cpp [hunk 9]
```diff
@@ -1131,6 +1225,152 @@ CCMD( say_team )
...
+CCMD( sayto_idx )
+{
+	int playerIndex = 0;
+
+	if ( argv.argc() >= 2 )
+	{
+		if ( argv.SafeGetNumber( 1, playerIndex ) == false )
+			return;
+
+		if ( PLAYER_IsValidPlayer( playerIndex ) == false )
```
Comment (TorrSamaho) [inline src/chat.cpp]: Should we allow the server to be contacted with idx "-1"?

File: src/chat.cpp [hunk 10]
```diff
@@ -1131,6 +1225,152 @@ CCMD( say_team )
...
+				g_ulChatPlayer = 0;
+
+			// [AK] If we're back to the old value, then there's nobody else to send a message to.
+			if ( g_ulChatPlayer == oldPlayer )
+			{
+				Printf( "There's no other player to send a private message to.\n" );
+				return false;
+			}
+		}
+		while (( PLAYER_IsValidPlayer( g_ulChatPlayer ) == false ) || ( g_ulChatPlayer == (ULONG)consoleplayer ));
```
Comment (TorrSamaho) [inline src/chat.cpp]: Looks like this will end in an infinite loop of there is only one player on the server.

File: src/chat.cpp [hunk 11]
```diff
@@ -1131,6 +1225,152 @@ CCMD( say_team )
 	}
 }
 
+//*****************************************************************************
+//
+// [AK] Allows players (or the server) to send private messages to other players.
+//
+bool chat_CanSendPrivateMessage( ULONG ulPlayer )
```
Comment (TorrSamaho) [inline src/chat.cpp]: The name of the function sounds as if it would only check whether `ulPlayer` is allowed to send a private message, but what it actually does it to change the global variable `g_ulChatPlayer` to a valid private chat target. The name of the function should reflect that. I would also rename the argument `ulPlayer` to `ulSender`.

File: src/chat.cpp [hunk 12]
```diff
@@ -1131,6 +1225,152 @@ CCMD( say_team )
...
+{
+	if ( ulPlayer == MAXPLAYERS )
```
Comment (TorrSamaho) [inline src/chat.cpp]: We don't need to make sure g_ulChatPlayer is a valid target if the sender is the server?

**Comment by TorrSamaho** [comment]
Sorry for overlooking the changes in `chat.cpp` last time. I didn't realize that github automatically hides changed files with too many changes...

File: src/chat.cpp (hunk 7 above)
Comment (TorrSamaho) [reply src/chat.cpp]: Good point. I agree that there is no need for private chat in single player or when watching demos.

**Comment by AKMDM** [comment]
No worries! I'm glad we took a good look in this file to see where things need to be improved for next time. I'll probably have these changes ready for next week, if that's okay. Thanks for the feedback!

File: src/chat.cpp (hunk 12 above)
Comment (AKMDM) [reply src/chat.cpp]: You're right, that needs to checked as well.

File: src/chat.cpp (hunk 9 above)
Comment (AKMDM) [reply src/chat.cpp]: Yes, this should be allowed too.

File: src/chat.cpp (hunk 8 above)
Comment (AKMDM) [reply src/chat.cpp:1387]: Technically yes, since it first assumes the client is sending a private message to the server rather than a player who may have the same name. I think it's unlikely somebody will try to use a name exactly like that (I remember some hosts would even forbid players from using it to prevent impersonation), but it's still possible. Should a player name exactly like "Server" even be allowed?

File: src/chat.cpp (hunk 6 above)
Comment (AKMDM) [reply src/chat.cpp]: I agree, I will add this check as well.

**Comment by TorrSamaho** [comment]
Sure, there's no rush. Have a nice week!

**Comment by AKMDM** [comment]
Thank you, you as well. :)

**Comment by AKMDM** [comment]
I submitted a couple of new commits.

- I fixed the private chat mode so that it now only works in online games and bots cannot be selected. Also, if the client tries to start typing a private message when there's no other clients, the server is selected instead.
- `chat_CanSendPrivateMessage( ULONG ulPlayer )` is replaced with `chat_IsPlayerValidReceiver( ULONG ulPlayer )` and `chat_FindValidReciever ( void )`. The former checks if a player is eligible to become a receiver of a private message, and the latter checks for a valid player that can be a receiver. Otherwise, the server is chosen.
- The possibility of any infinite loops occurring in the code should be resolved now.
- The server can now be selected by passing `-1` to `sayto_idx`.

I also added more commits, some of which aren't necessarily related to the private chat:

- Players cannot name themselves as "server" in online games anymore. This should avert potential impersonation of the server, and the possibility that someone tries using `sayto "server"` when there's a player with the same name.
- The chat cursor now blinks while the playing is creating a message. If they're typing or passing any input, the cursor remains visible. The additions are somewhat complicated, but I had to make sure `displayString` still looks convincing on the screen while the cursor is supposed to be invisible by taking account the width of the '_' character.
- I touched up and re-added my commit which adds the new EVENT type: `GAMEEVENT_CHAT` and the ACS function `GetChatMessage( int player, int index )`. The biggest difference is that up to three chat messages from the player (or server) are stored in memory, so modders can also pass an `index` between 0 and 2 to get that corresponding chat message, with ~~0 being the oldest and 2 being newest~~. This is in case we receive more than one chat message from the client in a single tic, in which the last chat message replaces the older one.

EDIT: An index of 0 returns the newest chat message, and 2 returns the oldest chat message.

File: src/chat.cpp [hunk 13]
```diff
@@ -508,6 +520,18 @@ void CHAT_Tick( void )
...
+
+	// [AK] Reset the chat cursor's ticker if it goes too high.
+	if (( g_ulChatMode != CHATMODE_NONE ) && ( ++g_ulChatTicker >= TICRATE ))
+		g_ulChatTicker = 0;
+
+	// [AK] If we're typing a chat message to another player, we must constantly check if
+	// they're in the game. If not, cancel the message entirely.
+	if ( g_ulChatMode == CHATMODE_PRIVATE_SEND )
+	{
+		if (( g_ulChatPlayer != MAXPLAYERS ) && ( PLAYER_IsValidPlayer( g_ulChatPlayer ) == false ))
```
Comment (TorrSamaho) [inline src/chat.cpp]: Shouldn't this should be using `chat_IsPlayerValidReceiver` instead of `PLAYER_IsValidPlayer` now?

File: src/chat.cpp [hunk 14]
```diff
@@ -594,7 +618,30 @@ bool CHAT_Input( event_t *pEvent )
 			}
 			else if ( pEvent->data1 == '\t' )
 			{
-				g_ChatBuffer.TabComplete();
+				if (( g_ulChatMode == CHATMODE_PRIVATE_SEND ) && ( SERVER_CountPlayers( false ) >= 2 ))
```
Comment (TorrSamaho) [inline src/chat.cpp:611]: This is also breaking now when `tempPlayer == MAXPLAYERS `, so there is no need anymore to check `SERVER_CountPlayers( false ) >= 2`, right?

Comment (AKMDM) [reply src/chat.cpp:611]: `SERVER_CountPlayers( false ) >= 2` is still needed, since there's no reason to execute the loop and try finding another client who we can privately chat with if none exist. However, the server (MAXPLAYERS) is always a valid receiver, so in this case, we don't really need to check if `tempPlayer` is a valid receiver using `chat_IsPlayerValidReceiver( tempPlayer )`. I hope this makes sense.

EDIT: Looking at it, I probably should've added an `if ( ulPlayer == MAXPLAYERS )` check to `chat_IsPlayerValidReceiver`, it could've simplified the code in a few places.

Comment (TorrSamaho) [reply src/chat.cpp:611]: Good point.

File: src/chat.cpp (hunk 13 above)
Comment (AKMDM) [reply src/chat.cpp]: I must've overlooked this line, sorry about that.

**Comment by TorrSamaho** [comment]
Thanks for the new patches! However, to keep things more manageable, can you do me a favor and remove the separate commits cd80940, d2fc101, f976178 from this merge request? Let's first get the private chat system merged, before extending it or adding other chat features.

**Comment by AKMDM** [comment]
Sure, will do.

~~EDIT: Is it okay if I open up a new pull request? I'm having trouble removing those few commits on my end and ended up making some mistakes~~.

**Comment by AKMDM** [comment]
I managed to fix it, Torr.

File: src/chat.cpp [hunk 15]
```diff
@@ -1131,6 +1226,215 @@ CCMD( say_team )
...
+		g_ulChatPlayer = MAXPLAYERS;
+	}
+}
+
+//*****************************************************************************
+//
+// [AK] Allows players (or the server) to send private messages to other players.
+//
+void chat_PrivateMessage( FCommandLine &argv, const ULONG ulReceiver, const bool bServerSelected )
+{
```
Comment (TorrSamaho) [inline src/chat.cpp:1298]: Why do we need a separate argument `bServerSelected` to encode the server as target instead of just using `ulReceiver == MAXPLAYERS`?
BTW: Sorry that this comment only shows up now. I posted it last week but unfortunately using "Start a review" instead of "Add single comment". I just now realized that you couldn't see it.

**Comment by TorrSamaho** [comment]
> I managed to fix it, Torr.

Thanks! Looking all good now, except for the one comment I didn't properly post last week. Sorry for that.

File: src/chat.cpp (hunk 15 above)
Comment (AKMDM) [reply src/chat.cpp:1298]: Thank you for asking. I added `bServerSelected` to ignore all the extra checks required to validate another player being the receiver, if the client explicitly chooses to the server. If the client wants indicates that they want to send a private message to the server using `sayto server` or `sayto_idx -1`, then it's not necessary to check if the receiver is invalid, a bot, ourselves, etc.

Additionally, if `bServerSelected` is true, then the function `chat_FindValidReceiver` won't be entered at all. If we did enter that function and there was only one client connected, then the message "There's no other clients to send private messages to. The server will be selected instead." would always appear in the console, even if the client intentionally chose to send a private message to the server.

Comment (TorrSamaho) [reply src/chat.cpp:1298]: You could also bypass all those checks by using `ulReceiver == MAXPLAYERS` instead of `bServerSelected`. For instance, you could add `const bool bServerSelected = (ulReceiver == MAXPLAYERS);` in the first line of the implementation of `chat_PrivateMessage` and keep the rest of `chat_PrivateMessage` untouched (except for removing the argument `bServerSelected`). Perhaps there is a reason why you don't want to encode the information `bServerSelected` in the argument `ulReceiver`, but right now I don't see it.

Comment (AKMDM) [reply src/chat.cpp:1298]: Fair enough, I can change that quickly and resubmit it if you're still around.

Comment (TorrSamaho) [reply src/chat.cpp:1298]: Yes, I'm still around.

Comment (TorrSamaho) [reply src/chat.cpp:1298]: Sorry, but I gotta go now. I should be able to add this tomorrow though.

Comment (AKMDM) [reply src/chat.cpp:1298]: No worries, sorry for the wait. I just around to adding that last patch, but if you can still add it in tomorrow that's awesome!

File: src/chat.cpp [hunk 16]
```diff
@@ -1294,10 +1294,11 @@ void chat_FindValidReceiver( void )
 //
 // [AK] Allows players (or the server) to send private messages to other players.
 //
-void chat_PrivateMessage( FCommandLine &argv, const ULONG ulReceiver, const bool bServerSelected )
+void chat_PrivateMessage( FCommandLine &argv, const ULONG ulReceiver )
 {
 	ULONG		ulIdx;
 	FString		ChatString;
+	const bool	bServerSelected = ulReceiver;
```
Comment (TorrSamaho) [inline src/chat.cpp]: Shouldn't this be `const bool bServerSelected = (ulReceiver == MAXPLAYERS);`?

EDIT: Now I'm really gone for today ;).

Comment (AKMDM) [reply src/chat.cpp]: Yes, I forgot to fix that (really stupid mistake). :(

EDIT: Uploaded the now fixed commit, so sorry about that!

Comment (TorrSamaho) [reply src/chat.cpp]: Thanks!

**Comment by TorrSamaho** [comment]
I compressed, rebased and added this patch.
